ibis/expr/operations.py

```diff
@@ -2353,3 +2353,63 @@ class TimestampDelta(ValueOp):
 
     input_type = [rules.timestamp, rules.timedelta(name='offset')]
     output_type = rules.shape_like_arg(0, 'timestamp')
+
+
+class ArrayLength(ValueOp):
+
+    input_type = [rules.array_array(dt.any)]
+    output_type = rules.shape_like_arg(0, 'int64')
+
+
+class ArraySlice(ValueOp):
+
+    input_type = [
+        rules.array_array(dt.any),
+        rules.integer(name='start'),
+        rules.integer(name='stop')
+    ]
+    output_type = rules.array_output(lambda self: self.args[0].type())
+
+
+class ArrayIndex(ValueOp):
+
+    input_type = [rules.array_array(dt.any), rules.integer(name='index')]
+    output_type = rules.array_output(
+        lambda self: self.args[0].type().value_type
+    )
+
+
+def _array_concat_output_type(self):
+    """Find the leaf type of each of the left and right arrays in a concat
+    operation and build an output type respecting the nesting level of the
+    original array.
+
+    Notes
+    -----
+    What do other systems do when casting ARRAY<T> to ARRAY<U> where
+    CAST(T AS U) is a valid operation? Postgres allows this for numeric types.
```

Comment:
Reword this, because as it stands there's nothing special about this. My original intent was questioning whether any backends exhibit type variance with complex types. For example, can you use an `ARRAY<T>` wherever you can use an `ARRAY<U>` if `T` can be *implicitly* cast to `U`. I'm unaware of any SQL system that behaves this way.